File: code/data_structures/priority_queue/MinDHeap.py


class MinDHeap:
    def __init__(self, degree, maxNodes):
        self.d = max(2, degree)
        self.n = max(self.d, maxNodes)

        self.heap = [None] * self.n
        self.child = [0] * self.n
        self.parent = [0] * self.n
        for i in range(self.n):
            self.parent[i] = (i - 1) // self.d
            self.child[i] = i * self.d + 1

        self.sz = 0

    def size(self):
        return self.sz

    def isEmpty(self):
        return self.sz == 0

    def clear(self):
        self.heap = [None] * self.n
        self.sz = 0

    def peek(self):
        if self.isEmpty():
            return None
        return self.heap[0]

    def poll(self):
        if self.isEmpty():
            return None
        root = self.heap[0]
        self.heap[0] = self.heap[self.sz - 1]
        self.heap[self.sz - 1] = None
        self.sz -= 1
        self.sink(0)
        return root

    def add(self, elem):
        if elem is None:
            raise ValueError("No null elements please :)")
        self.heap[self.sz] = elem
        self.swim(self.sz)
        self.sz += 1

    def sink(self, i):
        while True:
            j = self.minChild(i)
            if j == -1:
                break
     

File: code/data_structures/priority_queue/MinDHeap.cpp


#include <vector>
#include <stdexcept>

template<typename T>
class MinDHeap {
private:
    std::vector<T> heap;
    std::vector<int> child, parent;
    int d, n, sz;

public:
    MinDHeap(int degree, int maxNodes) : d(std::max(2, degree)), n(std::max(d, maxNodes)), heap(n), child(n), parent(n), sz(0) {
        for (int i = 0; i < n; ++i) {
            parent[i] = (i - 1) / d;
            child[i] = i * d + 1;
        }
    }

    int size() {
        return sz;
    }

    bool isEmpty() {
        return sz == 0;
    }

    void clear() {
        heap.clear();
        child.clear();
        parent.clear();
        sz = 0;
    }

    T peek() {
        if (isEmpty()) return T();
        return heap[0];
    }

    T poll() {
        if (isEmpty()) return T();
        T root = heap[0];
        heap[0] = heap[--sz];
        heap[sz] = T();
        sink(0);
        return root;
    }

    void add(const T& elem) {
        if (elem == nullptr) throw std::invalid_argument("No null elements please :)");
        heap[sz] = elem;
        swim(sz);
        ++sz;
    }

private:
    void sink(int i) {
        int j = minChild(i);
        while (j != -1) {
            swap(i, j);
            i = j;
            j = minChild(i);
        }
    }

    void swim(int i) {
        while (less(i, parent[i])) {
            swap(i, parent[i]);
            i = parent[i];
        }
    }

    int minChild(int i) {
        int index = -1;
        int from = child[i];
        int to = std::min(sz, from + d);
        for (int j = from; j < to; ++j)
            if (less(j, i)) {
                index = i = j;
            }
        return index;
    }

    bool less(int i, int j) {
        return heap[i] < heap[j];
    }

    void swap(int i, int j) {
        T tmp = heap[i];
        heap[i] = heap[j];
        heap[j] = tmp;
    }
};
